File: posixlib/src/main/resources/scala-native/wordexp.c
#if defined(SCALANATIVE_COMPILE_ALWAYS) || defined(__SCALANATIVE_POSIX_WORDEXP)
#if defined(__unix__) || defined(__unix) || defined(unix) ||                   \
    (defined(__APPLE__) && defined(__MACH__))
#ifndef __OpenBSD__

#include <stddef.h>
#include <wordexp.h>

struct scalanative_wordexp_t {
    size_t we_wordc; //  Count of words matched by 'words'.
    char **we_wordv; // Pointer to list of expanded words.
    size_t we_offs;  // Slots to reserve at the beginning of we_wordv.

    /* Permitted but not requited by POSIX 2018.
     * Used here to allow direct overlay calling on FreeBSD in addition
     * to Linux & macOS.
     */
    char *we_strings;  // storage for wordv strings
    size_t *we_nbytes; // size of we_strings
};

#if !(defined __STDC_VERSION__) || (__STDC_VERSION__ < 201112L)
#ifndef SCALANATIVE_SUPPRESS_STRUCT_CHECK_WARNING
#warning "Size and order of C structures are not checked when -std < c11."
#endif
#else

_Static_assert(sizeof(struct scalanative_wordexp_t) >= sizeof(wordexp_t),
               "size mismatch: wordexp_t");

_Static_assert(offsetof(struct scalanative_wordexp_t, we_wordc) ==
                   offsetof(wordexp_t, we_wordc),
               "offset mismatch: wordexp_t we_wordc");

_Static_assert(offsetof(struct scalanative_wordexp_t, we_wordv) ==
                   offsetof(wordexp_t, we_wordv),
               "offset mismatch: wordexp_t we_wordv");

_Static_assert(offsetof(struct scalanative_wordexp_t, we_offs) ==
                   offsetof(wordexp_t, we_offs),
               "offset mismatch: wordexp_t we_offs");

#endif // __STDC_VERSION__
#endif // !OpenBSD
#endif // Unix or Mac OS

// flags
#if !defined(__OpenBSD__)
int scalanative_wrde_append() { return WRDE_APPEND; };

int scalanative_wrde_dooffs() { return WRDE_DOOFFS; };

int scalanative_wrde_nocmd() { return WRDE_NOCMD; };

int scalanative_wrde_reuse() { return WRDE_REUSE; };

int scalanative_wrde_showerr() { return WRDE_SHOWERR; };

int scalanative_wrde_undef() { return WRDE_UNDEF; };

// error returns
int scalanative_wrde_badchar() { return WRDE_BADCHAR; };

int scalanative_wrde_badval() { return WRDE_BADVAL; };

int scalanative_wrde_cmdsub() { return WRDE_CMDSUB; };

int scalanative_wrde_nospace() { return WRDE_NOSPACE; };

int scalanative_wrde_syntax() { return WRDE_SYNTAX; };

#endif // not bsd

#endif // __SCALANATIVE_POSIX_WORDEXP

File: posixlib/src/main/scala/scala/scalanative/posix/wordexp.scala
package scala.scalanative
package posix

import scalanative.unsafe._
import scalanative.unsafe.Nat._

import scalanative.posix.sys.types.size_t

/** POSIX wordexp.h for Scala
 *
 *  The Open Group Base Specifications
 *  [[https://pubs.opengroup.org/onlinepubs/9699919799 Issue 7, 2018]] edition.
 */

@extern
object wordexp {

  type wordexp_t = CStruct5[
    size_t, //  we_wordc  Count of words matched by 'words'.
    Ptr[CString], // we_wordv  Pointer to list of expanded words.
    size_t, // we_offs   Slots to reserve at the beginning of we_wordv.

    /* Permitted but not required by POSIX 2018.
     * Used here to allow direct overlay calling on FreeBSD in addition
     * to Linux & macOS.
     */
    Ptr[CString], // we_strings, storage for wordv strings
    size_t // we_nbytes, size of we_strings
  ]

  /// Symbolic constants
  // flags

  @name("scalanative_wrde_append")
  def WRDE_APPEND: CInt = extern

  @name("scalanative_wrde_dooffs")
  def WRDE_DOOFFS: CInt = extern

  @name("scalanative_wrde_nocmd")
  def WRDE_NOCMD: CInt = extern

  @name("scalanative_wrde_reuse")
  def WRDE_REUSE: CInt = extern

  @name("scalanative_wrde_showerr")
  def WRDE_SHOWERR: CInt = extern

  @name("scalanative_wrde_undef")
  def WRDE_UNDEF: CInt = extern

  // error returns
  @name("scalanative_wrde_badchar")
  def WRDE_BADCHAR: CInt = extern

  @name("scalanative_wrde_badval")
  def WRDE_BADVAL: CInt = extern

  @name("scalanative_wrde_cmdsub")
  def WRDE_CMDSUB: CInt = extern

  @name("scalanative_wrde_nospace")
  def WRDE_NOSPACE: CInt = extern

  @name("scalanative_wrde_syntax")
  def WRDE_SYNTAX: CInt = extern

  /// Methods

  def wordexp(
      pattern: CString,
      expansion: Ptr[wordexp_t],
      flags: CInt
  ): CInt = extern

  def wordfree(wordexpP: Ptr[wordexp_t]): CInt = extern
}

object wordexpOps {
  import wordexp.wordexp_t

  implicit class wordexp_tOps(val ptr: Ptr[wordexp_t]) extends AnyVal {
    def we_wordc: size_t = ptr._1
    def we_wordv: Ptr[CString] = ptr._2
    def we_offs: size_t = ptr._3
    // FreeBSD POSIX extensions
    def we_strings: Ptr[CString] = ptr._4
    def we_nbytes: size_t = ptr._5

    def we_wordc_=(v: size_t): Unit = ptr._1 = v
    def we_wordv_=(v: Ptr[CString]): Unit = ptr._2 = v
    def we_offs_=(v: size_t): Unit = ptr._3 = v
    // FreeBSD POSIX extensions
    def we_strings_=(v: Ptr[CString]): Unit = ptr._4
    def we_nbytes_=(v: size_t): Unit = ptr._5 = v
  }
}
